---
title: "Using UMI4Cats to analyze UMI-4C data"
author: "Mireia Ramos RodrÃ­guez & Marc Subirana-Granes"
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document:
    fig_width: 6
    fig_height: 4
vignette: >
  %\VignetteIndexEntry{Using UMI4Cats to analyze UMI-4C data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction
Hello stranger! If you are here, that means you've successfully completed the UMI-4C protocol and got some sequencing results.

```{r}
devtools::load_all('/imppc/labs/lplab/share/marc/repos/umi4cats')
```

# Quick start
```{r, eval=F}
# setup python dependencies

umi4catsVenvInst()

# generate umi4c counts

raw_dir <- system.file("raw_dir", package = "UMI4Cats")
wk_dir <- system.file("wk_dir", package = "UMI4Cats")
threads <- 1
bait_seq <- 'CCCAAATCGCCCAGACCAG'
bait_pad <- 'GCGCGGATC'
res_e <- "GATC"
cut_pos <- 0 
ref_gen <- '/imppc/labs/lplab/share/marc/refgen/hg19/hg19.fa'
genomic_track <- '/imppc/labs/lplab/share/marc/refgen/hg19/genomicTracksR/dpnII_genomicTrack.tsv'


umi4CatsContacts(raw_dir,
                 wk_dir,
                 threads,
                 bait_seq,
                 bait_pad,
                 res_e,
                 ref_gen,
                 genomic_track)


```

```{r}

## Analyzing single profiles -------------------

```

```{r, eval=F}

## Comparison of two profiles ------------------

```



# Processing UMI-4C data

UMI4Cats package requires some python dependencies. For this reason is necessary create a pyhton virtualenv with all the required modules. The required modules are defined in the python/requirements.txt. To ensure proper functioning the virtual environment should be named and placed in python/umi4catsVenv. How ever, is possible to automatically generate this virtual environment using the umi4catsVenvInst() function:

```{r}

# Install and set up all python dependencies
umi4catsVenvInst()

```

Genomic track generation

For analysing the fastq UMI4c data is necessary a genome track of the desired restriction enzyme. Using genomicTrackGenerator() function is possible to generate this genome track. The function is based on BSgenome and the name of the desired reference genome have to be in BSgenome format as the function description sets. 

```{r}

library(UMI4Cats)

output <- '/imppc/labs/lplab/share/marc/refgen/genomicTracksR'

#the selected RE in this case is DpnII (|GATC), so the cs5p is "" and cs3p is GATC
cs5p <- ""
cs3p <- "GATC"
nameRe <- 'dpnII'
refgen <- "BSgenome.Hsapiens.UCSC.hg19"
genomicTrackGenerator <- function(cp5p, cs3p, nameRe, refgen, output)
  
```


# Analyzing UMI-4C data
Once the processing step is finished, you should end up with a text file containing the following informtion:
- chr bait
- start bait
- chr contact
- start contact
- number of UMIs

This file needs to be loaded into R and then processed

## Analysis of single profiles

The first function one need to use to analyze the processed UMI-4C results in `processUMI4C`. The input of this function can be either a character with the filename where the result of the processing is stored or an already loaded data.frame with that information. 

The input data.frame needs to have 5 columns:

- Bait coordinates (Columns 1:2): `chr` and `start`.
- Contact coordinates (Columns 3:4): `chr` and `start`.
- Number of UMIs supporting the contact (Column 5). 

The `processUMI4C` function will automatically obtain the bait coordinates from the input data.frame. If you want to provide a custom name for the bait that will show in the plot, you need to use the argument `bait_name`.

If for some reason you need to specify custom bait coordinates, you can do so by using the `bait_coordinates` argument. This argument should be a `GRanges` object with the coordinates of the custom bait and one `mcols` element representing the name for the bait (if you want to leave it empty, just add the mcols with value `""`).  



```{r, fig.width=5, fig.height=3}
ctrl.proc <- processUMI4C("../data-raw/B0M6_CCR1-5_TFRC_enhUmiRecount.txt",
                          factor=0.01)

plot <- plotUMI4C(ctrl.proc)
plot
```


- `processUMI4C`
- `plotUMI4C`

## Comparative analysis of profiles
- `processComparisonUMI4C`
- `plotComparisonUMI4C`
- `differentialContacts`

Demultiplex samples 

UMI4Cats 
```{r}
barcodeM <- c('KLK3',	'ATGGTCTGGGCGCTGTCTTG',
              'KLK6',	'TATTCTTCCTCAGCCCACATCTT',
              'KLK7',	'GGATGAAGATTTTGGAGCCCAGC',
              'KLK10',	'GGGCGGGGATTGAACGC')

barcodeM <- t(matrix(barcodeM, nrow = 2))
dfBarcodes <- data.frame(barcodeM)

colnames(dfBarcodes) <- c('sample', 'barcode')
fastqFile <- '/home/labs/lplab/msubirana/Documents/IGTP/kallikreins/raw/fastq/umi4c/CTTGTA_R1.fastq.gz'
type <- 'paired'
outPath <- '/home/labs/lplab/msubirana/Documents/IGTP/kallikreins/raw/fastq'

fastqMultxR(dfBarcodes,
            fastqFile,
            type,
            outPath)
```
