---
title: "Using UMI4Cats to analyze UMI-4C data"
author: "Mireia Ramos-RodrÃ­guez & Marc Subirana-Granes"
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document:
    fig_width: 6
    fig_height: 4
vignette: >
  %\VignetteIndexEntry{Using UMI4Cats to analyze UMI-4C data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction
Hello stranger! If you are here, that means you've successfully completed the UMI-4C protocol and got some sequencing results.

```{r}
library(UMI4Cats)
```

# Quick start
```{r processing-quick-start, eval=FALSE}
## Bait info
bait_seq <- 'ACCTAGAAGGATATGCGCTTGC'
bait_pad <- 'GCGTTAGA'
res_e <- "GATC"
cut_pos <- 0 

## Working directories
wd <- system.file("extdata/raw_dir", package = "UMI4Cats")
raw_dir <- file.path(wd, "extdata/raw_dir")
wk_dir <- "workdir"

## Tools
fastqmultx <- 'fastq-multx'
bowtie2 <- 'bowtie2'
samtools <- 'samtools'
py_umi4cats <- '/imppc/labs/lplab/share/env/general/bin/python3'

## Needed files
ref_gen <- '/imppc/labs/lplab/share/marc/refgen/hg19/hg19.fa'
genomic_track <- file.path(wd, "extdata/genomicTracks/dpnII_genomicTrack")

umi4CatsContacts(py_umi4cats,
                 raw_dir,
                 wk_dir,
                 threads,
                 bait_seq,
                 bait_pad,
                 res_e,
                 ref_gen,
                 genomic_track,
                 trimmomatic,
                 bowtie2,
                 fastqmultx,
                 samtools)
```

```{r analysis-quick-start}

```


# Processing UMI-4C data


# Analyzing UMI-4C data
Once the processing step is finished, you should end up with a text file containing the following information:
- chr bait
- start bait
- chr contact
- start contact
- number of UMIs

