<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analyzing UMI-4C data with UMI4Cats • UMI4Cats</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analyzing UMI-4C data with UMI4Cats">
<meta property="og:description" content="">
<meta property="og:image" content="/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">UMI4Cats</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/umi4cats-usage.html">Analyzing UMI-4C data with UMI4Cats</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Analyzing UMI-4C data with UMI4Cats</h1>
                        <h4 class="author">Mireia Ramos-Rodríguez &amp; Marc Subirana-Granés</h4>
            
            <h4 class="date">2020-01-08</h4>
      
      
      <div class="hidden name"><code>umi4cats-usage.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<div class="figure">
<img src="../man/figures/logo.png">
</div>
<p>Hello stranger! If you are here, that means you’ve successfully completed the UMI-4C protocol and got some sequencing results.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(UMI4Cats)</code></pre></div>
</div>
<div id="quick-start" class="section level1">
<h1 class="hasAnchor">
<a href="#quick-start" class="anchor"></a>Quick start</h1>
<div id="processing-fastq-files" class="section level2">
<h2 class="hasAnchor">
<a href="#processing-fastq-files" class="anchor"></a>Processing FASTQ files</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## 1) Generate Digested genome ----------------------------
<span class="co"># The selected RE in this case is DpnII (|GATC), so the cs5p is "" and cs3p is GATC</span>
hg19_dpnii &lt;-<span class="st"> </span><span class="kw"><a href="../reference/digestGenome.html">digestGenome</a></span>(<span class="dt">cut_pos =</span> <span class="dv">0</span>,
                           <span class="dt">res_enz =</span> <span class="st">"GATC"</span>,
                           <span class="dt">name_RE =</span> <span class="st">"DpnII"</span>,
                           <span class="dt">refgen =</span> BSgenome.Hsapiens.UCSC.hg19<span class="op">::</span>BSgenome.Hsapiens.UCSC.hg19, 
                           <span class="dt">out_path =</span> <span class="st">"digested_genome/"</span>)

## 2) Process UMI-4C fastq files --------------------------
raw_dir &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="st">"extdata"</span>, <span class="st">"SOCS1"</span>, <span class="st">"fastq"</span>), 
                       <span class="dt">package=</span><span class="st">"UMI4Cats"</span>)

<span class="kw"><a href="../reference/contactsUMI4C.html">contactsUMI4C</a></span>(<span class="dt">fastq_dir =</span> raw_dir,
              <span class="dt">wk_dir =</span> <span class="st">"SOCS1"</span>,
              <span class="dt">bait_seq =</span> <span class="st">"CCCAAATCGCCCAGACCAG"</span>,
              <span class="dt">bait_pad =</span> <span class="st">"GCGCG"</span>,
              <span class="dt">res_enz =</span> <span class="st">"GATC"</span>,
              <span class="dt">cut_pos =</span> <span class="dv">0</span>,
              <span class="dt">digested_genome =</span> hg19_dpnii,
              <span class="dt">ref_gen =</span> <span class="st">"/biodata/indices/species/Hsapiens/ucsc.hg19.fa"</span>,
              <span class="dt">threads =</span> <span class="dv">5</span>)</code></pre></div>
</div>
<div id="analyzing-umi-4c-results" class="section level2">
<h2 class="hasAnchor">
<a href="#analyzing-umi-4c-results" class="anchor"></a>Analyzing UMI-4C results</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Get filtering and alignment stats ---------
<span class="kw"><a href="../reference/statsUMI4C.html">statsUMI4C</a></span>(<span class="dt">wk_dir =</span> <span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"SOCS1"</span>,
                               <span class="dt">package=</span><span class="st">"UMI4Cats"</span>))</code></pre></div>
<p><img src="umi4cats-usage_files/figure-html/analysis-quick-start-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
<span class="co"># Load sample processed file paths</span>
files &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"SOCS1"</span>, <span class="st">"count"</span>, 
                                <span class="dt">package=</span><span class="st">"UMI4Cats"</span>),
                    <span class="dt">pattern=</span><span class="st">"*_altCounts.tsv"</span>,
                    <span class="dt">full.names=</span><span class="ot">TRUE</span>)

<span class="co"># Create colData including all relevant information</span>
colData &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="dt">sampleID =</span> <span class="kw"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">"_altCounts.tsv"</span>, <span class="st">""</span>, <span class="kw"><a href="https://rdrr.io/r/base/basename.html">basename</a></span>(files)),
                      <span class="dt">file =</span> files,
                      <span class="dt">stringsAsFactors=</span>F)

<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(tidyr)
colData &lt;-<span class="st"> </span>colData <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://tidyr.tidyverse.org/reference/separate.html">separate</a></span>(sampleID, 
           <span class="dt">into=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"condition"</span>, <span class="st">"replicate"</span>, <span class="st">"viewpoint"</span>),
           <span class="dt">remove=</span><span class="ot">FALSE</span>)

<span class="co"># Load UMI-4C data and generate UMI4C object</span>
umi &lt;-<span class="st"> </span><span class="kw"><a href="../reference/UMI4C.html">makeUMI4C</a></span>(<span class="dt">colData=</span>colData,
                 <span class="dt">viewpoint_name=</span><span class="st">"SOCS1"</span>)

<span class="co"># Perform differential test</span>
umi &lt;-<span class="st"> </span><span class="kw"><a href="../reference/fisherUMI4C.html">fisherUMI4C</a></span>(umi,
                   <span class="dt">filter_low =</span> <span class="dv">20</span>)

<span class="co"># Plot results</span>
<span class="kw"><a href="../reference/plotUMI4C.html">plotUMI4C</a></span>(umi, 
          <span class="dt">ylim=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>,<span class="dv">10</span>),
          <span class="dt">xlim=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">11e6</span>, <span class="fl">11.5e6</span>))</code></pre></div>
<p><img src="umi4cats-usage_files/figure-html/analysis-quick-start-2.png" width="672"></p>
</div>
</div>
<div id="experiment-design" class="section level1">
<h1 class="hasAnchor">
<a href="#experiment-design" class="anchor"></a>Experiment design</h1>
<!-- TODO: Explain how FastQ files should be organized for projects with different viewpoints -->
</div>
<div id="processing-umi-4c-fastq-files" class="section level1">
<h1 class="hasAnchor">
<a href="#processing-umi-4c-fastq-files" class="anchor"></a>Processing UMI-4C FASTQ files</h1>
<div id="reference-genome-digestion" class="section level2">
<h2 class="hasAnchor">
<a href="#reference-genome-digestion" class="anchor"></a>Reference genome digestion</h2>
<p>For the processing of the UMI-4C FastQ files it is necessary to construct a digested genome using the appropriate restriction enzyme. The <code>UMI4Cats</code> package includes the <code><a href="../reference/digestGenome.html">digestGenome()</a></code> function to make this process as easy as possible. The function uses a <code>BSgenome</code> object<a href="#fn1" class="footnoteRef" id="fnref1"><sup>1</sup></a> as reference genome and given a cutting sequence for the description enzyme, digests the reference genome.</p>
<p>The cutting position is defined by <code>cut_pos</code> a zero-based numeric integrer indicating the nucleotide position where restriction enzyme cuts.</p>
<p>In the following table you can see two examples of the different cutting sequences for <em>DpnII</em>, <em>Csp6I</em> and <em>HindIII</em>.</p>
<table class="table">
<thead><tr class="header">
<th>Restriction enzyme</th>
<th>Restriction seq</th>
<th>cut_pos</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>DpnII</td>
<td><code>|GATC</code></td>
<td>0</td>
</tr>
<tr class="even">
<td>Csp6I</td>
<td><code>G|TAC</code></td>
<td>1</td>
</tr>
<tr class="odd">
<td>HindIII</td>
<td><code>A|AGCTT</code></td>
<td>1</td>
</tr>
</tbody>
</table>
<p>For this example, we are using the hg19 <code>BSGenome</code> object and we are going to digest it using <em>DpnII</em>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(BSgenome.Hsapiens.UCSC.hg19)
refgen &lt;-<span class="st"> </span>BSgenome.Hsapiens.UCSC.hg19

hg19_dpnii &lt;-<span class="st"> </span><span class="kw"><a href="../reference/digestGenome.html">digestGenome</a></span>(<span class="dt">res_enz =</span> <span class="st">"GATC"</span>,
                           <span class="dt">cut_pos =</span> <span class="dv">0</span>,
                           <span class="dt">name_RE =</span> <span class="st">"dpnII"</span>,
                           <span class="dt">ref_gen =</span> refgen, 
                           <span class="dt">out_path =</span> <span class="st">"digested_genome/"</span>)

hg19_dpnii
<span class="co">#&gt; [1] "digested_genome//BSgenome.Hsapiens.UCSC.hg19_dpnII"</span>
<span class="co">#&gt; [1] "digested_genome/BSgenome.Hsapiens.UCSC.hg19_DpnII.tsv"</span></code></pre></div>
<p>Additionally, the <code><a href="../reference/digestGenome.html">digestGenome()</a></code> function returns invisibly the path of the generated file, so you can redirect it to a variable (in this case <code>hg19_dpnii</code>).</p>
</div>
<div id="demultiplex-fastq-files-for-different-baits" class="section level2">
<h2 class="hasAnchor">
<a href="#demultiplex-fastq-files-for-different-baits" class="anchor"></a>Demultiplex FastQ files for different baits</h2>
<p>UMI4Cats package provides a fastq-multx parser for demultiplexing paired-end FastQ files. Barcodes and bait names need to be provided as a data.frame as follows in the next example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">barcodes &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="dt">sample=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"KLK3"</span>, <span class="st">"KLK6"</span>, <span class="st">"KLK7"</span>, <span class="st">"KLK10"</span>),
                      <span class="dt">barcode=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"ATGGTCTGGGCGCTGTCTTG"</span>,
                                <span class="st">"TATTCTTCCTCAGCCCACATCTT"</span>,
                                <span class="st">"GGATGAAGATTTTGGAGCCCAGC"</span>,
                                <span class="st">"GGGCGGGGATTGAACGC"</span>))

<span class="kw"><a href="../reference/demultiplexFastq.html">demultiplexFastq</a></span>(<span class="dt">barcodes=</span>barcodes,
                <span class="dt">fastq=</span><span class="st">"~/samples/sample_1_R1.fastq.gz"</span>,
                <span class="dt">type=</span><span class="st">"paired"</span>,
                <span class="dt">out_path=</span><span class="st">"raw_fastq"</span>)</code></pre></div>
</div>
<div id="umi-4c-fastq-files-processing" class="section level2">
<h2 class="hasAnchor">
<a href="#umi-4c-fastq-files-processing" class="anchor"></a>UMI-4C FastQ files processing</h2>
<p>This step converts paired-end reads in the FastQ files to UMI counts in the genome fragments generated using the restriction enzymes. The basic function for doing this analysis is <code><a href="../reference/contactsUMI4C.html">contactsUMI4C()</a></code> and should be ran in samples generated with the same viewpoint and restriction enzyme.</p>
<p>The function will consider all FastQ files in the same folder <code>fastq_dir</code> to be part of the same experiment (viewpoint + restriction enzyme). The paired-end FASTQ files in this folder have to have one of the follow extensions: <code>_RX.fastq</code>, <code>_RX.fq</code>, <code>_RX.fq.gz</code> or <code>_RX.fastq.gz</code>.</p>
<p>For every experiment analysis, the user needs to define 3 different sequences: - <strong>Bait/viewpoint sequence</strong> (<code>bait_seq</code>). This is the downstream primer sequence (DS primer) that matches the sequence of the queried bait. - <strong>Padding sequence</strong> (<code>bait_pad</code>). The padding sequence corresponds to the nucleotides betweeen the DS primer end and the restriction enzyme site. - <strong>Restriction enzyme sequence</strong> (<code>res_enz</code>). This sequence is the restriction sequence recognized by the selected restriction enzyme.</p>
<div class="figure">
<img src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fnmeth.3922/MediaObjects/41592_2016_Article_BFnmeth3922_Fig1_HTML.jpg?as=webp">
</div>
<p>Additionally, it is necessary to define the restiction enzyme cutting postion (<code>cut_pos</code>) in the restriction sequence which corresponds to the site where the restriction enzyme cuts the sequence. For example, for DpnII (<code>|GATC</code>) <code>cut_pos = 0</code> and for Csp6I (<code>G|TAC</code>) <code>cut_pos = 1</code>. The function will also need the path for the digested genome (<code>digested_genome</code>) generated with the function <code><a href="../reference/digestGenome.html">digestGenome()</a></code>.</p>
<p>As <code><a href="../reference/contactsUMI4C.html">contactsUMI4C()</a></code> performs the alignment using <a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">Bowtie2</a>, you will also need to provide a reference genome in fasta format and the corresponding indexes generated with bowtie2<a href="#fn2" class="footnoteRef" id="fnref2"><sup>2</sup></a>, which need to be located in the same folder as the fasta file.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">raw_dir &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="st">"extdata"</span>, <span class="st">"SOCS1"</span>, <span class="st">"fastq"</span>), 
                       <span class="dt">package=</span><span class="st">"UMI4Cats"</span>)

<span class="kw"><a href="../reference/contactsUMI4C.html">contactsUMI4C</a></span>(<span class="dt">fastq_dir =</span> raw_dir,
              <span class="dt">wk_dir =</span> <span class="st">"SOCS1"</span>,
              <span class="dt">bait_seq =</span> <span class="st">"CCCAAATCGCCCAGACCAG"</span>,
              <span class="dt">bait_pad =</span> <span class="st">"GCGCG"</span>,
              <span class="dt">res_enz =</span> <span class="st">"GATC"</span>,
              <span class="dt">cut_pos =</span> <span class="dv">0</span>,
              <span class="dt">digested_genome =</span> hg19_dpnii,
              <span class="dt">ref_gen =</span> <span class="st">"/biodata/indices/species/Hsapiens/ucsc.hg19.fa"</span>,
              <span class="dt">threads =</span> <span class="dv">5</span>)</code></pre></div>
<p><code><a href="../reference/contactsUMI4C.html">contactsUMI4C()</a></code> will run the following processes sequentially:</p>
<ol style="list-style-type: decimal">
<li>FastQ files preparation (<code>prepUMI4C</code>).</li>
<li>Split reads at restriction sequences (<code>splitUMI4C</code>).</li>
<li>Align split reads to the reference genome (<code>alignmentUMI4C</code>).</li>
<li>Algorithm for collapsing reads coming from same molecule, returning all fragment ends 10Mb around the viewpoint(<code>counterUMI4C</code>).</li>
</ol>
<p>For the preparation and splitting, the FastQ file is loaded into memory. If you are having problems with the memory size, you can change the number of lines to be loaded using the <code>numb_reads</code> parameter. See <code><a href="../reference/contactsUMI4C.html">?contactsUMI4C</a></code> for more information.</p>
<p>Finally, the output (UMI-4C counts) is a tsv file saved in <code>wk_dir/count</code>. This file will be the input for the data analysis in the following section.</p>
</div>
<div id="quality-control-measures" class="section level2">
<h2 class="hasAnchor">
<a href="#quality-control-measures" class="anchor"></a>Quality control measures</h2>
<p>Once the contact analysis has been ran, the statistics of the UMI-4C filtering, aligment and final number of UMIs can be generated from the logs generated by the <code><a href="../reference/contactsUMI4C.html">contactsUMI4C()</a></code> function.</p>
<p>By using these logs, the function <code><a href="../reference/statsUMI4C.html">statsUMI4C()</a></code> will produce a summary plot .</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/statsUMI4C.html">statsUMI4C</a></span>(<span class="dt">wk_dir =</span> <span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"SOCS1"</span>,
                                <span class="dt">package=</span><span class="st">"UMI4Cats"</span>))</code></pre></div>
<p><img src="umi4cats-usage_files/figure-html/unnamed-chunk-5-1.png" width="672"></p>
</div>
</div>
<div id="analyzing-umi-4c-data" class="section level1">
<h1 class="hasAnchor">
<a href="#analyzing-umi-4c-data" class="anchor"></a>Analyzing UMI-4C data</h1>
<p>For this part of the analysis each bait should be analyzed separately. A different <code>UMI4C</code> object should be created for each analyzed bait.</p>
<div id="construct-the-umi4c-object" class="section level2">
<h2 class="hasAnchor">
<a href="#construct-the-umi4c-object" class="anchor"></a>Construct the <code>UMI4C</code> object</h2>
<p>The first step of the UMI-4C data analysis consists on loading the tables generated by the function <code>umi4CatsContacts</code> and use them to construct a <code>UMI4C</code>object. All these steps are perform automatically by the <code>makeUMI4C</code> functions.</p>
<p>The <code>makeUMI4C</code> will need as imput a data frame (<code>colData</code>) containing all relevant experiment information that will be needed for analyzing the data later on. The mandatory columns that this data frame should have are:</p>
<ol style="list-style-type: decimal">
<li>
<code>sampleID</code>: Unique identifier for the sample.</li>
<li>
<code>replicate</code>: Replicate identifier or number.</li>
<li>
<code>condition</code>: Condition for performing the diferential analysis. For example: “control” and “treatment”, two different cell types, etc. The condition column should only have <strong>two</strong> different values. If more condition variables are provided, the differential analysis will fail.</li>
<li>
<code>file</code>: Complete path and filename where the tsv files generated by <code>umi4CatsContacts</code> are saved.</li>
</ol>
<p>You can also include additional columns to <code>colData</code>, but the ones mentioned above are mandatory.</p>
<p>Additionally, the <code>makeUMI4C</code> function also contains other arguments that can be used if you want to tweak the default parameters of the analysis. See <code><a href="../reference/UMI4C.html">?makeUMI4C</a></code> to have a complete list and description of all the arguments.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Load sample processed file paths</span>
files &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"SOCS1"</span>, <span class="st">"count"</span>, 
                                <span class="dt">package =</span> <span class="st">"UMI4Cats"</span>),
                    <span class="dt">pattern =</span> <span class="st">"*_altCounts.tsv"</span>,
                    <span class="dt">full.names =</span> <span class="ot">TRUE</span>)

<span class="co"># Create colData including all relevant information</span>
colData &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="dt">sampleID =</span> <span class="kw"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">"_altCounts.tsv"</span>, <span class="st">""</span>, <span class="kw"><a href="https://rdrr.io/r/base/basename.html">basename</a></span>(files)),
                      <span class="dt">file =</span> files,
                      <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)

<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(tidyr)
colData &lt;-<span class="st"> </span>colData <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://tidyr.tidyverse.org/reference/separate.html">separate</a></span>(sampleID, 
           <span class="dt">into =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"condition"</span>, <span class="st">"replicate"</span>, <span class="st">"viewpoint"</span>),
           <span class="dt">remove =</span> <span class="ot">FALSE</span>)

<span class="co"># Load UMI-4C data and generate UMI4C object</span>
umi &lt;-<span class="st"> </span><span class="kw"><a href="../reference/UMI4C.html">makeUMI4C</a></span>(<span class="dt">colData =</span> colData,
                 <span class="dt">viewpoint_name =</span> <span class="st">"SOCS1"</span>)

umi
<span class="co">#&gt; class: UMI4C </span>
<span class="co">#&gt; dim: 2895 4 </span>
<span class="co">#&gt; metadata(5): bait scales min_win_factor region ref_umi4c</span>
<span class="co">#&gt; assays(6): umis norm_mat ... scale sd</span>
<span class="co">#&gt; rownames(2895): frag_26371 frag_26372 ... frag_29268 frag_29269</span>
<span class="co">#&gt; rowData names(2): id_contact position</span>
<span class="co">#&gt; colnames(4): ctrl_hi19_SOCS1 ctrl_hi22_SOCS1 cyt_hi19_SOCS1</span>
<span class="co">#&gt;   cyt_hi22_SOCS1</span>
<span class="co">#&gt; colData names(5): sampleID condition replicate viewpoint file</span></code></pre></div>
<p>This function contains several basic steps for preparing and analzying the UMI-4C data:</p>
<ol style="list-style-type: decimal">
<li>Removing fragment ends around the bait. The default value can be changed using the <code>bait_exclusion</code> argument.</li>
<li>Focusing the scope of the analysis in a specific region around the bait. The default value can be changed using the <code>bait_expansion</code> argument.</li>
<li>Obtaining the normalization matrices that will be used to scale the samples to the reference sample. If you want to avoid this normalization step, you can set <code>normalized</code> to <code>FALSE</code>.</li>
<li>Calculating the domainograms for each sample.</li>
<li>Calculating the adaptative trend for each sample.</li>
</ol>
</div>
<div id="performing-differential-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#performing-differential-analysis" class="anchor"></a>Performing differential analysis</h2>
<p>Once the <code>UMI4C</code> object is generated, one can perform a differential analysis using the <code><a href="../reference/fisherUMI4C.html">fisherUMI4C()</a></code> function. This will take as input the <code>UMI4C</code> object. If no other argument is provided, it will bin the region around the viewpoint and perform Fisher’s Exact tests in those windows, comparing the number of reads in each condition at that specific window with the number of reads in the studied region. Finally, p-values will be corrected using the FDR method and regions with adjusted p-values &gt; 0.05 will be considered signficantly different. The <code>filter_low</code> arguments determines which windows are tested: if they do not contain at least the number of UMIs in <code>filter_low</code>, they will not be tested. If you want to disable the filtering you can set <code>filter_low=FALSE</code>. You can also change the size of the windows used for this analysis using the <code>window_size</code> argument. Check <code>?fisherUMI4()</code> for more information.</p>
<p>Additionally, one can select the regions of interest to query for differential contacts using the <code>query_regions</code> argument.</p>
<p>Results from this test can be retrieved using the <code>results()</code> method defined for <code>UMI4C</code> objects.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">umi_dif &lt;-<span class="st"> </span><span class="kw"><a href="../reference/fisherUMI4C.html">fisherUMI4C</a></span>(umi,
                   <span class="dt">filter_low=</span><span class="dv">30</span>)

<span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw">results</span>(umi_dif))
<span class="co">#&gt; GRanges object with 6 ranges and 6 metadata columns:</span>
<span class="co">#&gt;       seqnames            ranges strand |          id    pvalue odds_ratio</span>
<span class="co">#&gt;          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt;  &lt;numeric&gt;</span>
<span class="co">#&gt;   [1]    chr16 10849721-10854720      * |    region_1      &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt;   [2]    chr16 10854721-10859720      * |    region_2      &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt;   [3]    chr16 10859721-10864720      * |    region_3      &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt;   [4]    chr16 10864721-10869720      * |    region_4      &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt;   [5]    chr16 10869721-10874720      * |    region_5      &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt;   [6]    chr16 10874721-10879720      * |    region_6      &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt;       log2_odds_ratio      padj      sign</span>
<span class="co">#&gt;             &lt;numeric&gt; &lt;numeric&gt; &lt;logical&gt;</span>
<span class="co">#&gt;   [1]            &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span>
<span class="co">#&gt;   [2]            &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span>
<span class="co">#&gt;   [3]            &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span>
<span class="co">#&gt;   [4]            &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span>
<span class="co">#&gt;   [5]            &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span>
<span class="co">#&gt;   [6]            &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></code></pre></div>
</div>
<div id="plotting-the-umi4c-object" class="section level2">
<h2 class="hasAnchor">
<a href="#plotting-the-umi4c-object" class="anchor"></a>Plotting the <code>UMI4C</code> object</h2>
<p>Once the <code>UMI4C</code> object is created, you can plot it using the <code>plotUMI4C</code> function. By default, the function will combine the samples using the <strong>condition</strong> column from <code>colData(umi)</code> (the colData data frame you provided for generating the <code>UMI4C</code> object), but you can change this behaviour using the <code>grouping</code> argument. You can give it one or several column names and will to the grouping based on those variables. For example, if you set <code>grouping="sampleID"</code> it will plot the individual profiles. You can group the data using several columns as well.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotUMI4C.html">plotUMI4C</a></span>(umi)</code></pre></div>
<p><img src="umi4cats-usage_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>If the <code>UMI4C</code> object contains information on the differential contacts, this data will be shown in the plot as well.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotUMI4C.html">plotUMI4C</a></span>(umi_dif)</code></pre></div>
<p><img src="umi4cats-usage_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>More information on <code>BSgenome</code> package and objects can be found <a href="https://bioconductor.org/packages/release/bioc/html/BSgenome.html">here</a>)<a href="#fnref1">↩</a></p></li>
<li id="fn2"><p>See the <a href="http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#getting-started-with-bowtie-2-lambda-phage-example">getting started section</a> on the Bowtie2 page for more information on how to generate the index for the reference genome.<a href="#fnref2">↩</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li>
<a href="#quick-start">Quick start</a><ul class="nav nav-pills nav-stacked">
<li><a href="#processing-fastq-files">Processing FASTQ files</a></li>
      <li><a href="#analyzing-umi-4c-results">Analyzing UMI-4C results</a></li>
      </ul>
</li>
      <li><a href="#experiment-design">Experiment design</a></li>
      <li>
<a href="#processing-umi-4c-fastq-files">Processing UMI-4C FASTQ files</a><ul class="nav nav-pills nav-stacked">
<li><a href="#reference-genome-digestion">Reference genome digestion</a></li>
      <li><a href="#demultiplex-fastq-files-for-different-baits">Demultiplex FastQ files for different baits</a></li>
      <li><a href="#umi-4c-fastq-files-processing">UMI-4C FastQ files processing</a></li>
      <li><a href="#quality-control-measures">Quality control measures</a></li>
      </ul>
</li>
      <li>
<a href="#analyzing-umi-4c-data">Analyzing UMI-4C data</a><ul class="nav nav-pills nav-stacked">
<li><a href="#construct-the-umi4c-object">Construct the <code>UMI4C</code> object</a></li>
      <li><a href="#performing-differential-analysis">Performing differential analysis</a></li>
      <li><a href="#plotting-the-umi4c-object">Plotting the <code>UMI4C</code> object</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Mireia Ramos-Rodriguez, Marc Subirana-Granes.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
