#'Umi4c analysis
#'@description
#' This function parses, clean, splits, aligns and analyses fastq files counting chromatin contacts between the viewpoint and
#'the surrounding loci. The output is a tsv file used for the generation of 4C profile.
#'
#'@usage #'umi4CatsContacts(pathVenv, raw_dir, wk_dir, threads, bait_seq, bait_pad, res_e, ref_gen, genomic_track)
#'
#'@param raw_dir Path with the fastq (compressed or descompressed)
#'@param threads Number of threads to use in the analysis
#'@param bait_seq Bait sequence
#'@param bait_pad Bait pad
#'@param res_e Restriction enzyme sequence
#'@param cut_pos Nucleotide where restriction enzyme cuts (zero-base)
#'@param ref_gen Reference genome in fasta format
#'@param wk_dir Working directory where to carry out the UMI-4c analysis
#'@param genomic_track Genomic track path generated by the generateGenomicTrack function of UMI4Cats package
#'@param pathVenv UMI4Cats virtual enviroment path
#'@examples
#'\dontrun{
#'pathVenv = '/imppc/labs/lplab/share/marc/venv/umi4catsVenv'
#'raw_dir = '/imppc/labs/lplab/share/marc/epimutations/processed/prove/raw'
#'wk_dir = '/imppc/labs/lplab/share/marc/epimutations/processed/prove/MLH1_ctrl_umi4cats_python'
#'threads = 1
#'bait_seq = 'ACCTAGAAGGATATGCGCTTGC'
#'bait_pad = 'GCGTTAGA'
#'res_e = "GATC"
#'cut_pos = "0" # should be a string for properly working
#'ref_gen = '/imppc/labs/lplab/share/marc/refgen/hg19/hg19.fa'
#'genomic_track = '/imppc/labs/lplab/share/marc/epimutations/processed/genomicTracks/genomic_tracks_hg19/dpnII_genomicTrack'
#'
#'umi4CatsContacts(pathVenv = pathVenv,
#'                 raw_dir = raw_dir,
#'                 wk_dir = wk_dir,
#'                 threads = threads,
#'                 bait_seq = bait_seq,
#'                 bait_pad = bait_pad,
#'                 res_e = res_e,
#'                 ref_gen = ref_gen,
#'                 genomic_track = genomic_track)
#'}
#' @export

umi4CatsContacts <- function(pathVenv,
                             raw_dir,
                             wk_dir,
                             threads,
                             bait_seq,
                             bait_pad,
                             res_e,
                             ref_gen,
                             genomic_track){

  prepR(pathVenv = pathVenv,
        raw_dir = raw_dir,
        wk_dir = wk_dir,
        bait_seq = bait_seq,
        bait_pad,
        res_e)

  splitR(pathVenv = pathVenv,
         wk_dir = wk_dir,
         res_e = res_e,
         cut_pos = cut_pos)

  alignmentR(pathVenv = pathVenv,
             wk_dir = wk_dir,
             threads = threads,
             ref_gen = ref_gen,
             bait_seq = bait_seq,
             bait_pad = bait_pad,
             res_e = res_e)

  umiCounterR(pathVenv = pathVenv,
              wk_dir = wk_dir,
              ref_gen = ref_gen,
              genomic_track = genomic_track,
              bait_seq = bait_seq,
              bait_pad = bait_pad,
              res_e = res_e)

  mergeUMICounter(genomic_track,
                  wk_dir)

}

