#'Umi4c analysis
#'@description
#' This function parses, clean, splits, aligns and analyses fastq files counting chromatin contacts between the viewpoint and 
#'the surrounding loci. The output is a tsv file used for the generation of 4C profile. 
#'  
#'@usage alignmentR(raw_dir, wk_dir, bait_seq, bait_pad, res_e, fastqmultx)
#'  
#'@param raw_dir Path with the fastq (compressed or descompressed)
#'@param threads Number of threads to use in the analysis
#'@param bait_seq Bait sequence
#'@param bait_pad Bait pad
#'@param res_e Restriction enzyme sequence
#'@param cut_pos Nucleotide where restriction enzyme cuts (zero-base)
#'@param ref_gen Reference genome in fasta format
#'@param wk_dir Working directory where to carry out the UMI-4c analysis
#'@param genomic_track Genomic track path
#'@param trimmomatic Trimmomatic path. By default trimmomatic
#'@param bowtie2 Bowtie2 path. By default bowtie2
#'@param fastqmultx Fastq-multx p√†th. By default fastq-multx
#'@param samtools Samtools path. By default samtools
#'@param genomic_track Genomic track path generated by the generateGenomicTrack function of UMI4Cats package
#'@param py_umi4cats
#'@examples
#'raw_dir = '/imppc/labs/lplab/share/marc/epimutations/processed/prove/raw'
#'wk_dir = '/imppc/labs/lplab/share/marc/epimutations/processed/prove/MLH1_ctrl_umi4cats_python'
#'threads = 1
#'bait_seq = 'ACCTAGAAGGATATGCGCTTGC'
#'bait_pad = 'GCGTTAGA'
#'res_e = "GATC"
#'cut_pos = "0" # should be a string for properly working
#'fastqmultx = 'fastq-multx'
#'bowtie2 = 'bowtie2'
#'ref_gen = '/imppc/labs/lplab/share/marc/refgen/hg19/hg19.fa'
#'samtools = 'samtools'
#'genomic_track = '/imppc/labs/lplab/share/marc/epimutations/processed/genomicTracks/genomic_tracks_hg19/dpnII_genomicTrack'
#'py_umi4cats = '/imppc/labs/lplab/share/marc/repos/umi4cats/python/umi4catsBuilder.py'

#'umi4CatsContacts(py_umi4cats,
#'                  raw_dir,
#'                  wk_dir,
#'                  threads,
#'                  bait_seq,
#'                  bait_pad,
#'                  res_e,
#'                  ref_gen,
#'                  genomic_track,
#'                  trimmomatic,
#'                  bowtie2,
#'                  fastqmultx,
#'                  samtools)
#'                  
#'@export

umi4CatsContacts <- function(py_umi4cats,
                             raw_dir,
                             wk_dir,
                             threads,
                             bait_seq,
                             bait_pad,
                             res_e,
                             ref_gen,
                             genomic_track,
                             trimmomatic = "trimmomatic",
                             bowtie2 = "bowtie2",
                             fastqmultx = "fastq-multx",
                             samtools = "samtools"){
  
  prepR(py_umi4cats,
        raw_dir,
        wk_dir,
        bait_seq,
        bait_pad,
        res_e,
        fastqmultx)
  
  splitR(py_umi4cats,
         wk_dir,
         res_e,
         cut_pos)
  
  alignmentR(py_umi4cats,
             wk_dir,
             threads,
             bowtie2,
             ref_gen,
             samtools,
             bait_seq,
             bait_pad,
             res_e)
  
  umiCounterR(py_umi4cats,
              wk_dir,
              bowtie2,
              ref_gen,
              samtools,
              genomic_track,
              bait_seq,
              bait_pad,
              res_e)
  
}

